File: sample.java
// ## Problem3 Candy (https://leetcode.com/problems/candy/)

// Time Complexity : O(m+n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : I didnt face any problem while coding this.


class Solution {
    public int candy(int[] ratings) {
        int[] result = new int[ratings.length];
        Arrays.fill(result, 1);
        int n = ratings.length;

        for(int i=1; i<n; i++){
            if(ratings[i] > ratings[i-1]){
                result[i] = result[i-1]+1;
            }
        }
        int min = result[n-1];
        for(int i=n-2; i>=0; i--){
            if(ratings[i] > ratings[i+1]){
                result[i] = Math.max(result[i], result[i+1] + 1);
            }
            min += result[i];
        }
        return min;
    }
}

// # Greedy-1

// ## Problem1 Jump Game (https://leetcode.com/problems/jump-game/)

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : I didnt face any problem while coding this.


class Solution {
    public boolean canJump(int[] nums) {
        int lastPosition = nums.length - 1;
        for(int i = nums.length - 1; i >= 0 ; i--){
            if(i + nums[i] >= lastPosition) {
                lastPosition = i;
            }
        }
        return lastPosition == 0;
    }
}

           
// ## Problem2 Jump Game II (https://leetcode.com/problems/jump-game-ii/)
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : I didnt face any problem while coding this.


class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        if(n == 1) return 0;
        int jumps = 1;
        int curr = nums[0];
        int next = nums[0];
        for(int i=0; i<nums.length; i++){
            next = Math.max(next, i+nums[i]);
            if(i== curr && i!= n-1){
                jumps++;
                curr = next;
            }
        }
        return jumps;
    }
}

File: sample.swift
// Time Complexity : 0(N)
// Space Complexity : 0(N)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

        if ratings.count == 1 {
            return 1
        }
        var candyarr = Array(repeating: 1, count: ratings.count)
        var candyarr2 = Array(repeating: 1, count: ratings.count)
        var count = 0
        for i in (0...ratings.count-2).reversed() {
            if ratings[i] > ratings[i+1]  {
                candyarr[i] = candyarr[i+1] + 1
            }
        }
        for i in 0..<candyarr2.count {
            if i != 0 && ratings[i] > ratings[i-1]  {
                candyarr2[i] = max(candyarr2[i-1] + 1,candyarr[i])
            } else {
                candyarr2[i] = candyarr[i]
            }
            count += candyarr2[i]
        }
        return count
    }
